src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -0,0 +1,1204 @@
+
+  mongoSubCacheItemInsert(cSubP);
+
+  return 0;
+}
+
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheItemInsert - create a new sub, fill it in, and add it to cache
+*/
+void mongoSubCacheItemInsert
+(
+  const char*               tenant,
+  const char*               servicePath,
+  SubscribeContextRequest*  scrP,
+  const char*               subscriptionId,
+  int64_t                   expirationTime,
+  int64_t                   throttling,
+  Format                    notifyFormat
+)
+{
+  CachedSubscription* cSubP = new CachedSubscription();
+  LM_T(LmtMongoSubCache,  ("allocated CachedSubscription at %p", cSubP));
+
+  //
+  // 1. First the non-complex values
+  //
+  cSubP->tenant                = (tenant[0] == 0)? NULL : strdup(tenant);
+  cSubP->servicePath           = strdup(servicePath);
+  cSubP->subscriptionId        = strdup(subscriptionId);
+  cSubP->reference             = strdup(scrP->reference.get().c_str());
+  cSubP->expirationTime        = expirationTime;
+  cSubP->throttling            = throttling;
+  cSubP->lastNotificationTime  = 0;
+  cSubP->pendingNotifications  = 0;
+  cSubP->notifyFormat          = notifyFormat;
+  cSubP->next                  = NULL;
+
+  LM_T(LmtMongoSubCache, ("inserting a new sub in cache (%s). lastNotifictionTime: %lu", cSubP->subscriptionId, cSubP->lastNotificationTime));
+
+
+  //
+  // 2. Then the values that have functions/methods for filling/parsing
+  //
+  cSubP->throttling = scrP->throttling.parse();
+  cSubP->notifyConditionVector.fill(scrP->notifyConditionVector);
+
+
+  //
+  // 3. Convert all patterned EntityId to EntityInfo
+  //
+  for (unsigned int ix = 0; ix < scrP->entityIdVector.vec.size(); ++ix)
+  {
+    EntityId* eIdP;
+
+    eIdP = scrP->entityIdVector.vec[ix];
+
+    if (eIdP->isPattern == "false")
+    {
+      continue;
+    }
+
+    EntityInfo* eP = new EntityInfo(eIdP->id, eIdP->type);
+    
+    cSubP->entityIdInfos.push_back(eP);
+  }
+
+
+  //
+  // 4. Insert all attributes 
+  //
+  for (unsigned int ix = 0; ix < scrP->attributeList.attributeV.size(); ++ix)
+  {
+    cSubP->attributes.push_back(scrP->attributeList.attributeV[ix]);
+  }
+
+
+  //
+  // 5. Now, insert the subscription in the cache
+  //
+  LM_T(LmtMongoSubCache, ("Inserting NEW sub '%s', lastNotificationTime: %lu", cSubP->subscriptionId, cSubP->lastNotificationTime));
+  mongoSubCacheItemInsert(cSubP);
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheStatisticsGet - 
+*/
+void mongoSubCacheStatisticsGet(int* refreshes, int* inserts, int* removes, int* updates, int* items, char* list, int listSize)
+{
+  *refreshes = mongoSubCache.noOfRefreshes;
+  *inserts   = mongoSubCache.noOfInserts;
+  *removes   = mongoSubCache.noOfRemoves;
+  *updates   = mongoSubCache.noOfUpdates;
+  *items     = mongoSubCacheItems();
+
+  CachedSubscription* cSubP = mongoSubCache.head;
+
+  //
+  // FIXME?
```

Comment:
Not really a FIXME, just a mark so that you would see it.
Please read and give me your opinion.
I've changed the "FIXME?" for a NOTE.

This information is **extremely** useful for debugging - as there are no other way to see the contents of the subscription cache right now.